File: frontend/src/lib/api/greenroom.ts
/**
 * Green Room API Client
 * Project Development & Voting Arena
 */

import { api } from '../api';

// ============ Types ============

export type CycleStatus = 'upcoming' | 'active' | 'closed';
export type ProjectStatus = 'pending' | 'approved' | 'rejected';
export type PaymentStatus = 'pending' | 'completed' | 'failed' | 'refunded';

export interface Cycle {
  id: number;
  name: string;
  description?: string;
  start_date: string;
  end_date: string;
  max_tickets_per_user: number;
  ticket_price: number;
  status: CycleStatus;
  created_at: string;
  updated_at: string;
  project_count?: number;
  total_votes?: number;
}

export interface Project {
  id: number;
  cycle_id: number;
  filmmaker_id: string;
  title: string;
  description: string;
  category?: string;
  video_url?: string;
  image_url?: string;
  status: ProjectStatus;
  vote_count: number;
  created_at: string;
  updated_at: string;
  approved_at?: string;
  filmmaker_name?: string;
  user_vote_count?: number;
}

export interface VotingTicket {
  id: number;
  user_id: string;
  cycle_id: number;
  tickets_purchased: number;
  tickets_used: number;
  tickets_available: number;
  payment_status: PaymentStatus;
  created_at: string;
}

export interface Vote {
  id: number;
  user_id: string;
  project_id: number;
  cycle_id: number;
  tickets_allocated: number;
  created_at: string;
  project_title?: string;
}

export interface CycleResults {
  cycle_id: number;
  cycle_name: string;
  status: CycleStatus;
  total_projects: number;
  total_votes: number;
  total_voters: number;
  projects: ProjectResult[];
}

export interface ProjectResult {
  project_id: number;
  title: string;
  filmmaker_id: string;
  filmmaker_name?: string;
  vote_count: number;
  rank: number;
}

export interface UserStats {
  total_tickets_purchased: number;
  total_tickets_used: number;
  total_votes_cast: number;
  projects_submitted: number;
  projects_approved: number;
}

export interface CycleStats {
  cycle_id: number;
  total_projects: number;
  approved_projects: number;
  pending_projects: number;
  rejected_projects: number;
  total_tickets_sold: number;
  total_votes_cast: number;
  unique_voters: number;
  revenue: number;
}

// ============ Request Types ============

export interface CycleCreateRequest {
  name: string;
  description?: string;
  start_date: string;
  end_date: string;
  max_tickets_per_user?: number;
  ticket_price?: number;
}

export interface CycleUpdateRequest {
  name?: string;
  description?: string;
  start_date?: string;
  end_date?: string;
  max_tickets_per_user?: number;
  ticket_price?: number;
  status?: CycleStatus;
}

export interface ProjectSubmitRequest {
  cycle_id: number;
  title: string;
  description: string;
  category?: string;
  video_url?: string;
  image_url?: string;
}

export interface ProjectUpdateRequest {
  title?: string;
  description?: string;
  category?: string;
  video_url?: string;
  image_url?: string;
}

export interface TicketPurchaseRequest {
  cycle_id: number;
  ticket_count: number;
}

export interface TicketPurchaseResponse {
  checkout_session_id: string;
  checkout_url: string;
  amount: number;
  ticket_count: number;
}

export interface VoteCastRequest {
  project_id: number;
  tickets_allocated: number;
}

// ============ API Client ============

export class GreenRoomAPI {
  private baseURL = '/api/v1/greenroom';

  // ============ Public Endpoints ============

  /**
   * List all cycles
   */
  async listCycles(status?: CycleStatus): Promise<Cycle[]> {
    const params = new URLSearchParams();
    if (status) params.append('status', status);

    return api.get(`${this.baseURL}/cycles?${params}`);
  }

  /**
   * Get cycle details
   */
  async getCycle(id: number): Promise<Cycle> {
    return api.get(`${this.baseURL}/cycles/${id}`);
  }

  /**
   * List projects in a cycle
   */
  async listProjects(
    cycleId: number,
    options?: {
      status?: ProjectStatus;
      sort_by?: 'votes' | 'recent' | 'title';
      skip?: number;
      limit?: number;
    }
  ): Promise<Project[]> {
    const params = new URLSearchParams();
    if (options?.status) params.append('status', options.status);
    if (options?.sort_by) params.append('sort_by', options.sort_by);
    if (options?.skip !== undefined) params.append('skip', options.skip.toString());
    if (options?.limit !== undefined) params.append('limit', options.limit.toString());

    return api.get(`${this.baseURL}/cycles/${cycleId}/projects?${params}`);
  }

  /**
   * Get project details
   */
  async getProject(id: number): Promise<Project> {
    return api.get(`${this.baseURL}/projects/${id}`);
  }

  /**
   * Get cycle results
   */
  async getCycleResults(cycleId: number): Promise<CycleResults> {
    return api.get(`${this.baseURL}/cycles/${cycleId}/results`);
  }

  // ============ User Endpoints ============

  /**
   * Get user's voting tickets
   */
  async getMyTickets(): Promise<VotingTicket[]> {
    return api.get(`${this.baseURL}/tickets/my-tickets`);
  }

  /**
   * Purchase voting tickets
   */
  async purchaseTickets(request: TicketPurchaseRequest): Promise<TicketPurchaseResponse> {
    return api.post(`${this.baseURL}/tickets/purchase`, request);
  }

  /**
   * Cast vote on a project (final, cannot be changed)
   */
  async castVote(vote: VoteCastRequest): Promise<Vote> {
    return api.post(`${this.baseURL}/votes/cast`, vote);
  }

  /**
   * Get user's votes
   */
  async getMyVotes(cycleId?: number): Promise<Vote[]> {
    const params = new URLSearchParams();
    if (cycleId) params.append('cycle_id', cycleId.toString());

    return api.get(`${this.baseURL}/votes/my-votes?${params}`);
  }

  /**
   * Get user statistics
   */
  async getMyStats(): Promise<UserStats> {
    return api.get(`${this.baseURL}/stats/my-stats`);
  }

  // ============ Filmmaker Endpoints ============

  /**
   * Submit a project
   */
  async submitProject(project: ProjectSubmitRequest): Promise<Project> {
    return api.post(`${this.baseURL}/projects/submit`, project);
  }

  /**
   * Get filmmaker's projects
   */
  async getMyProjects(): Promise<Project[]> {
    return api.get(`${this.baseURL}/projects/my-projects`);
  }

  /**
   * Update project (only if pending)
   */
  async updateProject(id: number, update: ProjectUpdateRequest): Promise<Project> {
    return api.put(`${this.baseURL}/projects/${id}`, update);
  }

  /**
   * Delete project (only if pending)
   */
  async deleteProject(id: number): Promise<void> {
    return api.delete(`${this.baseURL}/projects/${id}`);
  }

  // ============ Admin Endpoints ============

  /**
   * Create new cycle (admin only)
   */
  async createCycle(cycle: CycleCreateRequest): Promise<Cycle> {
    return api.post(`${this.baseURL}/cycles`, cycle);
  }

  /**
   * Update cycle (admin only)
   */
  async updateCycle(id: number, update: CycleUpdateRequest): Promise<Cycle> {
    return api.put(`${this.baseURL}/cycles/${id}`, update);
  }

  /**
   * Delete cycle (admin only)
   */
  async deleteCycle(id: number): Promise<void> {
    return api.delete(`${this.baseURL}/cycles/${id}`);
  }

  /**
   * Approve or reject project (admin only)
   */
  async approveProject(id: number, status: 'approved' | 'rejected'): Promise<Project> {
    return api.put(`${this.baseURL}/projects/${id}/approve`, { status });
  }

  /**
   * Get cycle statistics (admin only)
   */
  async getCycleStats(cycleId: number): Promise<CycleStats> {
    return api.get(`${this.baseURL}/cycles/${cycleId}/stats`);
  }
}

// Export singleton instance
export const greenroomAPI = new GreenRoomAPI();

File: backend/app/schemas/greenroom.py
"""
Green Room Pydantic Schemas
"""
from pydantic import BaseModel, Field, HttpUrl
from typing import Optional
from datetime import datetime
from app.models.greenroom import CycleStatus, ProjectStatus, PaymentStatus


# ============ Cycle Schemas ============

class CycleBase(BaseModel):
    """Base cycle schema"""
    name: str = Field(min_length=1, max_length=200)
    description: Optional[str] = None
    start_date: datetime
    end_date: datetime
    max_tickets_per_user: int = Field(default=100, ge=1, le=1000)
    ticket_price: float = Field(default=10.0, ge=0)


class CycleCreate(CycleBase):
    """Create new cycle"""
    pass


class CycleUpdate(BaseModel):
    """Update existing cycle"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    max_tickets_per_user: Optional[int] = Field(None, ge=1, le=1000)
    ticket_price: Optional[float] = Field(None, ge=0)
    status: Optional[CycleStatus] = None


class CycleResponse(CycleBase):
    """Cycle response"""
    id: int
    status: CycleStatus
    created_at: datetime
    updated_at: datetime
    project_count: Optional[int] = 0  # Total projects in cycle
    total_votes: Optional[int] = 0  # Total votes cast

    class Config:
        from_attributes = True


# ============ Project Schemas ============

class ProjectBase(BaseModel):
    """Base project schema"""
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=10)
    category: Optional[str] = Field(None, max_length=100)
    video_url: Optional[HttpUrl] = None
    image_url: Optional[HttpUrl] = None


class ProjectSubmit(ProjectBase):
    """Submit new project"""
    cycle_id: int


class ProjectUpdate(BaseModel):
    """Update project (only if pending)"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, min_length=10)
    category: Optional[str] = Field(None, max_length=100)
    video_url: Optional[HttpUrl] = None
    image_url: Optional[HttpUrl] = None


class ProjectResponse(ProjectBase):
    """Project response"""
    id: int
    cycle_id: int
    filmmaker_id: str
    status: ProjectStatus
    vote_count: int
    created_at: datetime
    updated_at: datetime
    approved_at: Optional[datetime] = None

    # Additional computed fields
    filmmaker_name: Optional[str] = None  # From User table
    user_vote_count: Optional[int] = None  # Current user's votes on this project

    class Config:
        from_attributes = True


class ProjectApproval(BaseModel):
    """Approve or reject project"""
    status: ProjectStatus = Field(..., description="approved or rejected")


# ============ Voting Ticket Schemas ============

class TicketPurchaseRequest(BaseModel):
    """Request to purchase voting tickets"""
    cycle_id: int
    ticket_count: int = Field(ge=1, le=100)


class TicketPurchaseResponse(BaseModel):
    """Response for ticket purchase (Stripe session)"""
    checkout_session_id: str
    checkout_url: str
    amount: float
    ticket_count: int


class VotingTicketResponse(BaseModel):
    """User's voting tickets for a cycle"""
    id: int
    user_id: str
    cycle_id: int
    tickets_purchased: int
    tickets_used: int
    tickets_available: int
    payment_status: PaymentStatus
    created_at: datetime

    class Config:
        from_attributes = True


# ============ Vote Schemas ============

class VoteCast(BaseModel):
    """Cast vote on a project"""
    project_id: int
    tickets_allocated: int = Field(ge=1, description="Number of tickets to allocate")


class VoteResponse(BaseModel):
    """Vote response"""
    id: int
    user_id: str
    project_id: int
    cycle_id: int
    tickets_allocated: int
    created_at: datetime

    # Additional fields
    project_title: Optional[str] = None

    class Config:
        from_attributes = True


# ============ Results Schemas ============

class ProjectResult(BaseModel):
    """Project with vote results"""
    project_id: int
    title: str
    filmmaker_id: str
    filmmaker_name: Optional[str] = None
    vote_count: int
    rank: int


class CycleResults(BaseModel):
    """Cycle voting results"""
    cycle_id: int
    cycle_name: str
    status: CycleStatus
    total_projects: int
    total_votes: int
    total_voters: int
    projects: list[ProjectResult]


# ============ Stats Schemas ============

class UserGreenRoomStats(BaseModel):
    """User's Green Room statistics"""
    total_tickets_purchased: int
    total_tickets_used: int
    total_votes_cast: int
    projects_submitted: int
    projects_approved: int


class CycleStats(BaseModel):
    """Cycle statistics"""
    cycle_id: int
    total_projects: int
    approved_projects: int
    pending_projects: int
    rejected_projects: int
    total_tickets_sold: int
    total_votes_cast: int
    unique_voters: int
    revenue: float  # Total ticket sales
